app/models/smart_proxy.rb

```diff
@@ -107,6 +108,44 @@ def version
     result
   end
 
+  def plugin_version(plugin_name)
+    raise ::Foreman::Exception.new(N_('Plugin name cannot be blank')) if plugin_name.blank?
+    result = {}
+    begin
+      Timeout::timeout(20) do
+        version = ProxyAPI::Version.new(:url => url).plugin_versions
+        result[:success] = true
+        result[:message] = version[plugin_name]
+      end
+    rescue *HTTP_ERRORS => exception
+      raise ::Foreman::WrappedException.new exception, N_("Unable to connect to smart proxy")
+    end
+    Rails.cache.write("plugin_#{plugin_name}_proxy_#{self.id}", result, {:expires_in  => 5.minutes })
+    result
+  end
+
+  def tftp_server
+    raise ::Foreman::Exception.new(N_('TFTP is not enabled for proxy: %s') % to_label) unless has_feature?('TFTP')
+    result = {}
+    begin
+      Timeout::timeout(20) do
+        server = ProxyAPI::TFTP.new(:url => url).bootServer
+        result[:success] = true
+        result[:message] = server
+      end
```

Comment:
These timeouts in model class do not look goot.

We we must use Ruby's Timeout, let's move it again into ProxyAPI and let's make this generic as I expect more plugins to use the same technique. Since there is a cached_get method already, this could become a decorator pattern now - we have cached queries and quieries with timeout. This is pseudocode:

    server = TimeoutAPI(CachedAPI(ProxyAPI::TFTP.new))

In ruby one can do decorators much more cleaner: http://nithinbekal.com/posts/ruby-decorators/